### ISSUE TITLE
Instantiating object only on form submission

### ISSUE THREAD

**Issue by ysrinu** [susanBuck/dwa15-fall2017#51]
In https://github.com/susanBuck/foobooks0/blob/master/getAllBooks.php
I think it makes sense to create a Book object only when the form is submitted. Would you agree?

This is similar to checking if $_GET is not empty before creating the Cipher object in https://github.com/susanBuck/dwa15-php-practice/blob/master/cipher/logic.php

The reason I am suggesting is because the class constructor might have code that would throw exception if the form has not been submitted yet (the code could be using constructor parameter values that are null). While this check can be done inside the constructor, the ideal approach would be to instantiate an object only when the form is submitted.

Going further, would it make sense that in the logic code (above 2 php files) do the 'is form submitted' check as the first thing and return immediately if false?

In other words,
instead of

```
if ($form->isSubmitted()) {
..... // 10s of lines
...
}
```

we could have

```
if (!$form->isSubmitted()) return;
... // 10s of lines
..

```

Thanks,
-Sri

**Comment by HeinrichTremblay** [comment]
That depends, for the Book object you might want to instantiate it for a default case if you haven't submit the form.

**Comment by SridharSubramaniam** [comment]
Yes, You would need it to load books to show on load, however, I  usually wouldn't preload the books in a constructor - rather would defer it to methods like getAllBooks - but check for null and load it only once. This would open up Book class to be used for other purposes before preloading. e.g if you really want to create a new book entry first before showing a list, you don't want to preload on instantiation, but by putting it in the constructor, we would end fetching all the time.  Another scenario would be, if you want to reload after create, we now do not have a method to refresh books data in a class and would be forced to create an another object.

Another point is, this books.php is more of a DAO that got operations on Book object and not an actual blueprint of Book itself - which would ideally be an another class with actual properties of book like title etc.

**Comment by mconstanti** [comment]
On your second point (if I'm understanding it correctly),  if the form is not submitted, the whole 'if' statement will be skipped. presumably going to the end of the file. So it would not be much different than returning immediately after checking if the `if (!$form->isSubmitted())` statement is true.

**Comment by susanBuck** [comment]
As currently written, correct - there's no reason we shouldn't check if we have form submission before instantiating the Book object. 

I believe this was an artifact from an earlier iteration of the app where we loaded books regardless of whether the form was submitted.

Refactored: https://github.com/susanBuck/foobooks0/commit/fe0cce12917ef08c41339f58425950c7de67217f

Regarding the styling of the if statement - if you prefer it the way you've described, go for it! I'd put this in the category of personal preference.

**Comment by ysrinu** [comment]
Thanks @susanBuck  and @mconstanti 
My statement regarding the if statement styling is indeed more about readability, to avoid big chunk of code indented to right just because of being inside the else clause. Removing the else cause (by returning at if (not) statement) makes it more readable, it doesn't make any difference in the performance though.
